### PR TITLE
Add test for a direct reference install

### PR DESCRIPTION
**Pull request by fridex**
## Related Issues and Dependencies

Related: #100
Fixes: #102 

## This introduces a breaking change

- [ ] Yes
- [x] No

## This Pull Request implements

Add a test for a direct dependency installation out of requirements.txt file. As the version string cannot be determined from the entry in requirements.txt file (none reliably provided), the installation process treats the stack as not fully locked.

### CODE REVIEW

**Comment by fridex** [comment]
CC @tumido

**Comment by fridex** [comment]
Looks like not all pip versions are compatible with such requirement. Marking as WIP.

**Comment by frenzymadness** [comment]
I don't understand the latest commit `Do not use @ in requirements.txt`. Why don't? [documentation of pip](https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers) have `@` between project name and `file://` URI and it works for me with it as well (for pip >= 19.3).

BTW, it seems that I've discovered a bug because the documentation says that the direct references are supported from pip version 19.1 but they're not. https://github.com/pypa/pip/pull/8604

Nitpicks: first commit contains debug print `print(artifact_path)` and the formatting change of unrelated code should be in a separated commit.

We'll need to find a way how to skip some tests for older pip versions.

**Comment by fridex** [comment]
> I don't understand the latest commit `Do not use @ in requirements.txt`. Why don't? [documentation of pip](https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers) have `@` between project name and `file://` URI and it works for me with it as well (for pip >= 19.3).
>
> BTW, it seems that I've discovered a bug because the documentation says that the direct references are supported from pip version 19.1 but they're not. [pypa/pip#8604](https://github.com/pypa/pip/pull/8604)

I wanted to trigger CI for different pip versions to see behavior. Exactly because of this. Thanks for the link.
 
> Nitpicks: first commit contains debug print `print(artifact_path)` and the formatting change of unrelated code should be in a separated commit.

Thanks, I'll squash commits.

> We'll need to find a way how to skip some tests for older pip versions.

I think this can be a solution provided by pytest:

https://docs.pytest.org/en/latest/skipping.html#id1

**Comment by frenzymadness** [comment]
> > We'll need to find a way how to skip some tests for older pip versions.
> 
> I think this can be a solution provided by pytest:
> 
> https://docs.pytest.org/en/latest/skipping.html#id1
> 
> Let's see if we can incorporate it with the venv fixture.

Definitely, but we have to get pip version somehow as a global variable. I'll take a look at how to do that.

**Comment by fridex** [comment]
> A possible solution:
> ...

I tried to experiment a bit with this - see https://github.com/thoth-station/micropipenv/pull/105. PTAL

File: tests/test_micropipenv.py [hunk 1]
```diff
@@ -253,6 +253,25 @@ def test_install_pip_tools_editable(venv):
             shutil.rmtree("micropipenv_editable_test.egg-info", ignore_errors=True)
 
 
+@pytest.mark.skipif(PIP_VERSION.release >= (19, 3, 0), reason="Direct reference installation is supported in pip starting 19.3")
```
Comment (frenzymadness) [inline tests/test_micropipenv.py:256]:
```suggestion
@pytest.mark.skipif(PIP_VERSION.release < (19, 3, 0), reason="Direct reference installation is supported in pip starting 19.3")
```

Comment (frenzymadness) [reply tests/test_micropipenv.py:256]: The test is skipped if the condition is True.

Comment (fridex) [reply tests/test_micropipenv.py]: Thanks :smiley:

**Comment by fridex** [comment]
Discarding the WIP flag. CI looks green. PTAL @frenzymadness

**Comment by fridex** [comment]
> The code looks good to me. The discussion in the pip docs PR uncovered that the feature this code tests might be also available in older pips than 19.3. Do we want to investigate it?

I'm open to adjust the check so we cover the older releases of pip as well in tests.

Following the discussion -- if the feature can be safely used in pip 19.3 (with the bug fix) I would consider that as a "starting release" where I could fully utilize direct references. Depends on the severity of bug that was fixed though.

**Comment by fridex** [comment]
I would do the condition adjustment in a separate PR if you are fine with that.

**Comment by frenzymadness** [comment]
> I would do the condition adjustment in a separate PR if you are fine with that.

I am. Feel free to merge this one.

**Pull request merged**